lighthouse-core/report/html/renderer/element-screenshot-renderer.js

```diff
@@ -0,0 +1,237 @@
+    containerEl.setAttribute('rectLeft', clipRect.left.toString());
+    containerEl.setAttribute('rectRight', clipRect.right.toString());
+    containerEl.setAttribute('rectTop', clipRect.top.toString());
+    containerEl.setAttribute('rectBottom', clipRect.bottom.toString());
+
+    // Zoom out (by double viewport but using 0.5 zoom factor) when highlighted region takes
+    // up most of the viewport. This provides more context for where on the page this element is.
+    /* todo: maybe only apply the width criterium in the preview screenshot (what's in the table, not the overlay) */
+    // if (clipRect.height > viewportSize.height / 2 || clipRect.width > viewportSize.width / 2) {
+    //   zoomFactor = 0.5;
+    //   viewport.width *= 2;
+    //   viewport.height *= 2;
+    // }
+
+    const intendedClipToViewportRatio = 0.75;
+    const zoomRatioXY = {
+      x: viewportSize.width / clipRect.width,
+      // x: 1, // ignore x?
+      y: viewportSize.height / clipRect.height,
+    };
+    const zoomRatio = intendedClipToViewportRatio * Math.min(zoomRatioXY.x, zoomRatioXY.y);
+    if (zoomRatio < 1) {
+      zoomFactor = zoomRatio;
+      viewport.width /= zoomRatio;
+      viewport.height /= zoomRatio;
+    }
+
+    viewport.width = Math.min(fullPageScreenshot.width, viewport.width);
+
+    const positions = ElementScreenshotRenderer.getScreenshotPositions(
+      clipRect,
+      viewport,
+      {width: fullPageScreenshot.width, height: fullPageScreenshot.height}
+    );
+
+    const contentEl = dom.find('.lh-element-screenshot__content', containerEl);
+    // TODO: can all the `* zoomFactor` be replaces with setting some CSS on the container?
+    // just `scale` doesn't work b/c won't change size of the element.
+    // containerEl.style.transform = `scale(${zoomFactor})`;
+    contentEl.style.top = `-${viewport.height * zoomFactor}px`;
+
+    const image = dom.find('.lh-element-screenshot__image', containerEl);
+    image.style.width = viewport.width * zoomFactor + 'px';
+    image.style.height = viewport.height * zoomFactor + 'px';
+
+    image.style.backgroundPositionY = -(positions.screenshot.top * zoomFactor) + 'px';
+    image.style.backgroundPositionX = -(positions.screenshot.left * zoomFactor) + 'px';
+    // image.style.backgroundSize = (zoomFactor * 100) + '%';
+    image.style.backgroundSize =
+      `${fullPageScreenshot.width * zoomFactor}px ${fullPageScreenshot.height * zoomFactor}px`;
+
+    const elMarker = dom.find('.lh-element-screenshot__element-marker', containerEl);
+    elMarker.style.width = clipRect.width * zoomFactor + 'px';
+    elMarker.style.height = clipRect.height * zoomFactor + 'px';
+    elMarker.style.left = positions.clip.left * zoomFactor + 'px';
+    elMarker.style.top = positions.clip.top * zoomFactor + 'px';
+
+    const mask = dom.find('.lh-element-screenshot__mask', containerEl);
+    const clipId = 'clip-' + Math.floor(Math.random() * 100000000);
```

Comment:
seems like we could do a sequentially increasing global clip id instead to be safe, the birthday problem is real

Reply:
updated, better?

Reply:
beautiful! could even make this whole thing internal to render clip path at that point :)